### PR TITLE
Added optional ETH phy thread for STM32 for link check

### PR DESCRIPTION
**Pull request by vlotech**
The ethernet driver of stm32 (low level driver) does not generate link information (cable attached or not). To allow applications to react on the link detect, an additional thread is added which poll the link status at a low rate. 
This can be switched on/off through the menuconfig.

Signed-off-by: Vincent van der Locht <vincent@vlotech.nl>

### CODE REVIEW

**Comment by pfalcon** [comment]
@vlotech, If this patch is for STM32, why phy_sam_gmac.h and friends are being patched here?

Also, did you write this from scratch, or followed pattern used in ethernet other driver(s)?

Thanks.

**Comment by pfalcon** [comment]
> and I did my best to align it with the mcux driver

Ok, that's what I meant, and good to know that you tried to be consistent with it.

> I can also add the atmel part as a whole to make it complete ?!?

Well, generally, that's discouraged - one PR, one separate change. Having too much stuff in one PR complicates the review, and of course makes it  longer - instead of waiting for review from just STM32 maintainer, you would need to wait from both STM32 and SAM.

With that warning, you can try to add SAM changes here, but please be ready that it will be requested to split.

> not experienced with that part of git yet

When you have time, please look up "git rebase interactive", that's universal way to deal with situation like this.


Thanks for the patch again.

**Comment by tbursztyka** [comment]
Just git reset --soft your_commit, git reset HEAD files_you_don't_want, and recommit the ones still in the list. Maybe there is an easier way if there is more than 1 commit to change.

That said, split your patch in logical changes: the ext/ stuff for instance should go first in there own patch. 

Also, adding one thread just for this is maybe too heavy, knowing that you already have a thread you can use: the one for rx processing. It would require a bit more work, but merging both rx/line-check in that one is I think the way to go. Also: that check would always be enabled then, no need for any Kconfig options then. (btw, stm32 does not provide any hw interrupt - gpio or else - for that kind of event?)

**Comment by vlotech** [comment]
I've force pushed the proposed update. Split between ext and zephyr driver code + removed atmel implementation.

File: ext/hal/st/stm32cube/stm32f1xx/drivers/include/stm32f1xx_hal_eth.h
Comment (erwango) [inline ext/hal/st/stm32cube/stm32f1xx/drivers/include/stm32f1xx_hal_eth.h]: You're not supposed to modify files in ext/hal/....
Whatever additional functionality you need should be implemented in drivers/ethernet/

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for the addition, though it is not possible to add new functionality under ext/hal/... packages.
Reason is that these packages are maintained externally.
Please reformat your changes to take this constraint into account.

File: ext/hal/st/stm32cube/stm32f1xx/drivers/include/stm32f1xx_hal_eth.h
Comment (vlotech) [reply ext/hal/st/stm32cube/stm32f1xx/drivers/include/stm32f1xx_hal_eth.h]: I've been looking for a way to implement the functionality solely under the driver part, not touching the ext/hal code. 

One way would be to partial or fully implement the HAL_ETH_Init(..) function in the driver/net/eth_stm32_hal.c codebase. Which doesn't speak for using the external low level drivers. 
Another way would be to call HAL_ETH_Init(..) every time the PHY state changes. There are two things why that wouldn't be ideal: 1) the init function needs to be called at boot anyhow to enable PHY communication and 2) more things than just the link status are being initialized. 
The init function has a long delay in case there's no link detected, which cannot be removed when leaving the ext/hal/.. code untouched and not reimplement it in the driver. 

I would like to ask the maintainers: @tbursztyka @erwango @pfalcon @jukkar what they would propose to overcome this long delay at boot.

Comment (erwango) [reply ext/hal/st/stm32cube/stm32f1xx/drivers/include/stm32f1xx_hal_eth.h]: @vlotech: checking for alternate and acceptable solution. Coming back to you.

File: drivers/ethernet/eth_stm32_hal.c
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c]: What actions do you actually need from this new function?

Comment (vlotech) [reply drivers/ethernet/eth_stm32_hal.c]: the HAL_ETH_CheckAutoNegotiation(..) function is the second part of the original HAL_ETH_Init(..) function which checks the actual PHY status and configure PHY, MAC and internal bookkeeping accordingly. 
So, the proposed solution is splitting the _init function in 2 parts. When the PHY thread is not used to actively check link status, the original code-progress is executed by first calling _init(..) and then _CheckAutoNegotiation(..). 
I added this split to touch the existing code as less as possible but still end up at the proposed end-situation.

**Comment by jukkar** [comment]
> Also, adding one thread just for this is maybe too heavy, knowing that you already have a thread you can use: the one for rx processing. It would require a bit more work, but merging both rx/line-check in that one is I think the way to go. Also: that check would always be enabled then, no need for any Kconfig options then. (btw, stm32 does not provide any hw interrupt - gpio or else - for that kind of event?)

Agreeing with @tbursztyka here, adding a thread is problematic as it requires separate stack which uses memory. One option that could perhaps be used is to use system work queue that could do the check periodically.

File: drivers/ethernet/eth_stm32_hal.c
Comment (erwango) [reply drivers/ethernet/eth_stm32_hal.c]: > checks the actual PHY status and configure PHY, MAC and internal bookkeeping accordingly

So, what is exactly required at this point and can't it be done thanks to existing functions like HAL_ETH_Read/WritePHYRegister, HAL_ETH_ConfigMAC, ... ?

**Comment by galak** [comment]
@vlotech any plans to update this PR?  If not we will close it and feel free to re-open when you are ready to resubmit.  Thanks.

**Comment by jukkar** [comment]
Carrier on/off support for stm32 ethernet was just merged in #19163 so I think this can be closed.

**Pull request closed without merging**
